File: jdba-core/src/main/java/org/jetbrains/jdba/core/DBFacade.java
package org.jetbrains.jdba.core;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.jdba.Rdbms;



/**
 * Represents a database.
 *
 * @author Leonid Bushuev from JetBrains
 */
public interface DBFacade extends ImplementationAccessibleService {


  /**
   * The DBMS this facade is applicable/connected to.
   */
  @NotNull
  Rdbms rdbms();


  /**
   * Connects to the database server.
   */
  void connect();

  /**
   * Reconnects from the database server.
   */
  void reconnect();

  /**
   * Disconnects from the database server.
   *
   * <p>
   *   If not connected - does nothing.
   * </p>
   */
  void disconnect();

  /**
   * Checks whether it is connected to the server.
   *
   * <p>
   *   TBD does it really performs a "ping" interaction or just returning a kind of internal status?
   * </p>
   *
   * @return whether is connected.
   */
  boolean isConnected();


  /**
   * Performs the given operation in transaction and returns the result.
   *
   * @param operation operation to perform.
   * @param <R>       type of result.
   * @return the result.
   * @see DBFacade#inTransaction
   */
  <R> R inTransaction(InTransaction<R> operation);

  /**
   * Performs the given operation in transaction and returns the result.
   *
   * @param operation operation to perform.
   * @see DBFacade#inTransaction
   */
  void inTransaction(InTransactionNoResult operation);


  <R> R inSession(InSession<R> operation);


  void inSession(InSessionNoResult operation);


}

File: dekaf-main/src/core/DBFacade.kt
package org.jetbrains.dekaf.core

import org.jetbrains.dekaf.Rdbms


/**
 * Represents a database.
 *
 * @author Leonid Bushuev
 */
interface DBFacade : DBSessionAware, ImplementationAccessibleService {


    /**
     * The DBMS this facade is applicable/connected to.
     */
    fun rdbms(): Rdbms

    /**
     * Activates the database driver.
     */
    fun activateDriver();

    /**
     * Connects to the database server.
     */
    fun connect()

    /**
     * Reconnects from the database server.
     */
    fun reconnect()

    /**
     * Ping the server to check whether it's alive.
     * @return respond delay, in ms.
     */
    fun ping(): Int

    /**
     * Disconnects from the database server.
     *
     * If not connected - does nothing.
     */
    fun disconnect()

    /**
     * Deactivates the driver. Also disconnects if necessary.
     */
    fun deactivateDriver();

    /**
     * Checks whether it is connected to the server.
     *
     * TBD does it really performs a "ping" interaction or just returning a kind of internal status?
     *
     * @return whether is connected.
     */
    val isConnected: Boolean

    /**
     * The driver information, when it's loaded and activated.
     * Null when is not activated.
     */
    val driverInfo: DbDriverInfo?

    /**
     * Provides a brief connection info.
     * @return a brief connection info.
     */
    val connectionInfo: ConnectionInfo


    /**
     * Lease a session.
     * When the session is not needed more, the [DBLeasedSession.close] must be called.
     *
     * @return a leased session.
     */
    fun leaseSession(): DBLeasedSession

}
